### ISSUE TITLE
formatoptions=j should ignore comment leader inside strings

### ISSUE THREAD

**Issue by luator** [neovim/neovim#14829]
- `nvim --version`:  Both v0.4.4 and v0.5.0-dev+1417-g6a77def1e
- Operating system/version: Ubuntu 18.04
- Terminal name/version: terminator
- `$TERM`: xterm-256color

*Update 2021-06-29: The title and the example below have been modified to focus on the actual issue.  See discussion below.*

### Steps to reproduce using `nvim -u NORC`

Enter the following text:
```
"Hello # World"
# foo
```
Make sure `formatoptions` contains `j` (should be the case by default). Then move to the first line and press `J`

### Actual behaviour
```
"Hello # World" foo
```

### Expected behaviour
```
"Hello # World" # foo
```

**Comment by chrisbra** [comment]
remove the `j` flag from your formatoptions.

**Comment by luator** [comment]
Oh, I see, that's indeed the cause.  When setting formatoptions explicitly, the behaviour is the same in neovim and vim.
I still consider it a bug, though, as the % in the upper line is inside a string and thus doesn't mark a comment.  But maybe this is beyond the parsing effort that is reasonable for this feature?

In any case the impact can be reduced by setting `comments` properly depending on the filetype (the above example is from Python where `%` is not a comment leader, so `:set comments=b:#` fixes this specific case for me).

**Comment by rktjmp** [comment]
The stock runtime files have what looks like the correct settings for python:

https://github.com/neovim/neovim/blob/master/runtime/ftplugin/python.vim#L38-L39

So you shouldn't need to patch in your own config.  If I test the above in a python file it works as it should.

You can run `:verbose set comments?` in a python file to see what might be clobbering the default value.

It's actually a bit odd to me that `%` is marked as a comment by default, or that comments _are_ defined for `ft=` but I presume that's just what vim did.

I guess if you run with ft detect off maybe you also want some defaults, *I guess...*? Seems like it would make more sense to default to `text`'s comment settings and if you want specific behaviour, you set the ft.

Anyway, seems like it's working as intended?

`:new`, `J` shows the "bug" behaviour
`:new`, `ft=python`, `J` shows the expected behaviour
`:new`, `ft=text`, `J` shows the expected behaviour.

**Comment by luator** [comment]
@rktjmp Thanks for pointing out that this should already be good by default.  I already tried setting filetype to Python before and it did not help in my case.  After searching a bit I found that I had some custom ftplugin file that set `b:did_ftplugin = 1` and thus prevented the global one from being loaded (I think this was some copy&paste from long ago...).  Removed this and now I get the correct behaviour without having to set `comments` manually :)

However, this also means that the ftplugins in `~/.config/nvim` are still loaded when using `nvim -u NORC` which is something I did not expect.  Is there an option to completely ignore any local configuration?

**Comment by chrisbra** [comment]
use `-u NONE`

**Comment by luator** [comment]
@chrisbra Thanks!  `-u NONE` seems to also prevent the global ftplugins from being used, so it is ignoring a bit too much.  But I realised that I can simply do something like

    export XDG_CONFIG_HOME=/tmp

before starting vim to make it ignore my local configuration.

**Comment by rktjmp** [comment]
@luator, do you still feel this is a bug or maybe it's closable?

**Comment by luator** [comment]
I still think it is somewhat a bug as the comment character in the first line is inside a string and thus does not really mark a comment.  E.g. in Python there would still be a problem with this example:
```
var = "Hello # world"
# comment
```
But I assume that fixing this in a generic way that works in all use cases might be unreasonably complicated (if possible at all), so I would be okay with closing it as won't fix.  I don't know the code, though, so maybe someone who does should decide?

**Comment by rktjmp** [comment]
PS: _Just to be clear, I am not saying you should close this, nor do I have any say in if it should, just there are 1k+ issues and sometimes dead ones hang around!_

That seems like a bug to me, or at least unfriendly/unexpected behaviour. Maybe we can change the title and first post to better reflect the issue though, or open a new one?

As you say, the actual issue is something like "joining a commented line onto a line with a comment marker inside a string should not remove the joined line's comment marker" but that's pretty verbose...

I think the desired behaviour might have to be limited to just ignoring comment markers inside `"`, `'` and maybe ` otherwise working out what is and isn't demarking a comment is probably more trouble than it's worth.

For example, in Elixir you can specify any string terminator you want, so `~s/a #string/` or `~sqa #stringq` and I think it behaving weirdly around that should just stay a known-ugly, considering the impact, otherwise where do you stop!

**Comment by luator** [comment]
I just updated title and first post.  Is the title clear like this?
I hope it isn't too confusing that now the issue description doesn't fully match with the initial discussion.

**Comment by rktjmp** [comment]
I tried a basic implementation here https://github.com/rktjmp/neovim/commit/5fa129c540a5ef4ca4a1460c2e8e9c527bbec559

But I don't think it is worth pursuing, you have to make some strong assumptions about what is and isn't a string terminator, `"` is pretty easy to do, but `'` requires you to check each instance isn't just punctuation inside a string. Or `"this \" weird string"`, etc.

To do it properly, you start heading down the road of writing an actual parser which is a mistake and I don't doubt there would be sneaky edge cases hiding around with different comment styles (multi lines) or languages that use `"` as a comment marker (though any language that does that is probably insane :suspect:). Adding the "fix" just opens the potential to *more* bugs IMO.

If enough people want it maybe it's worth it. I personally can't remember noticing the issue in years and years of use so :man_shrugging:, not to say that it isn't errant behaviour.

I think the *Real* way to fix it would be to run these kind of operations via treesitter, which means you just use an actual parser instead of writing a bad one. Not sure if TS would ever be that integrated though.

What you could do is remap `J` to use a TS query against the current line, if it has a comment on it, add `j` to your format options, otherwise remove it and then run the normal  `J`. That would probably be much more reliable.
